From 782ccf6397d446a89ce6fce19c027b5c267a33d8 Mon Sep 17 00:00:00 2001
From: Pau Ruiz Safont <pau.safont@vates.tech>
Date: Wed, 8 Oct 2025 15:14:57 +0100
Subject: [PATCH] xapi_vm_snapshot: move VDI-related DB operations inside
 revert_vbds

This moves them outside of the try catch that destroys newly-cloned
disks, but this is safe since these don't change the state of the host
like the destroy vifs.

This allows to simplify the type needed to be able to clean up the new
VDIs in case of failure.

Signed-off-by: Pau Ruiz Safont <pau.safont@vates.tech>
---
 ocaml/xapi/xapi_vm_snapshot.ml | 46 +++++++++++++++++-----------------
 1 file changed, 23 insertions(+), 23 deletions(-)

diff --git a/ocaml/xapi/xapi_vm_snapshot.ml b/ocaml/xapi/xapi_vm_snapshot.ml
index 743987c28..a4b491877 100644
--- a/ocaml/xapi/xapi_vm_snapshot.ml
+++ b/ocaml/xapi/xapi_vm_snapshot.ml
@@ -293,7 +293,24 @@ let revert_vbds ~__context ~rpc ~session_id ~snapshot ~vm =
         ~__context snap_suspend_VDI driver_params
   in
   TaskHelper.set_progress ~__context 0.6 ;
-  {disks= cloned_disks; cds= cloned_CDs; suspend_VDI= cloned_suspend_VDI}
+  debug "Copying the VBDs" ;
+  let (_ : [`VBD] Ref.t list) =
+    List.map
+      (fun (vbd, vdi, _) -> Xapi_vbd_helpers.copy ~__context ~vm ~vdi vbd)
+      (cloned_disks @ cloned_CDs)
+  in
+  debug "Update the suspend_VDI" ;
+  Db.VM.set_suspend_VDI ~__context ~self:vm ~value:cloned_suspend_VDI ;
+
+  cloned_suspend_VDI
+  :: List.fold_left
+       (fun acc (_, vdi, on_error_delete) ->
+         if on_error_delete then
+           vdi :: acc
+         else
+           acc
+       )
+       [] cloned_disks
 
 let update_vifs_vbds_vgpus_and_vusbs ~__context ~snapshot ~vm =
   let snap_VIFs = Db.VM.get_VIFs ~__context ~self:snapshot in
@@ -304,17 +321,11 @@ let update_vifs_vbds_vgpus_and_vusbs ~__context ~snapshot ~vm =
 
   (* clone all the disks of the snapshot *)
   Helpers.call_api_functions ~__context (fun rpc session_id ->
-      let cloned = revert_vbds ~__context ~rpc ~session_id ~snapshot ~vm in
+      let vdis_to_cleanup =
+        revert_vbds ~__context ~rpc ~session_id ~snapshot ~vm
+      in
+      TaskHelper.set_progress ~__context 0.7 ;
       try
-        debug "Copying the VBDs" ;
-        let (_ : [`VBD] Ref.t list) =
-          List.map
-            (fun (vbd, vdi, _) -> Xapi_vbd_helpers.copy ~__context ~vm ~vdi vbd)
-            (cloned.disks @ cloned.cds)
-        in
-        TaskHelper.set_progress ~__context 0.7 ;
-        debug "Update the suspend_VDI" ;
-        Db.VM.set_suspend_VDI ~__context ~self:vm ~value:cloned.suspend_VDI ;
         debug "Cleaning up the old VIFs" ;
         List.iter (safe_destroy_vif ~__context ~rpc ~session_id) vm_VIFs ;
         debug "Setting up the new VIFs" ;
@@ -341,18 +352,7 @@ let update_vifs_vbds_vgpus_and_vusbs ~__context ~snapshot ~vm =
         error
           "Error while updating the new VBD, VDI, VIF and VGPU records. \
            Cleaning up the cloned VDIs." ;
-        let vdis =
-          cloned.suspend_VDI
-          :: List.fold_left
-               (fun acc (_, vdi, on_error_delete) ->
-                 if on_error_delete then
-                   vdi :: acc
-                 else
-                   acc
-               )
-               [] cloned.disks
-        in
-        List.iter (safe_destroy_vdi ~__context ~rpc ~session_id) vdis ;
+        List.iter (safe_destroy_vdi ~__context ~rpc ~session_id) vdis_to_cleanup ;
         raise e
   )
 
